### PR TITLE
Refactor for labels & image outlier

### PR DESCRIPTION
**Pull request by matanper**
Before implementing label & image outliers checks, I needed a few changes:

1. easy way to get for the samples I run on the real dataset index. Will now be able to do in the update:
```
properties_with_id = []
for index, image in enumarte(batch.images):
  property = func(image)
  dataset_index = batch.get_index_in_dataset(index)
  properties_with_id += [(dataset_index, property)]
```
2. Made label & prediction properties to either return flat list or list of lists
3. Made it possible to run checks on VisionData classes that don't have labels. Until now it would have raised exception when trying to update the VisionData cache, so added try/except

### CODE REVIEW

**Review by noamzbr — APPROVED** [review]

File: tests/vision/base/test_base_checks.py [hunk 1]
```diff
@@ -28,13 +28,20 @@
...
-from deepchecks.core.errors import DatasetValidationError
+
 from deepchecks.vision.base_checks import SingleDatasetCheck, TrainTestCheck, ModelOnlyCheck
-from deepchecks.vision import ClassificationData, DetectionData
+from deepchecks.vision import ClassificationData, DetectionData, Context, Batch
 from deepchecks.vision.datasets.detection import coco
-from deepchecks.vision.vision_data import TaskType, VisionData
+
+
+def run_update_loop(dataset):
```
Comment (JKL98ISR) [inline tests/vision/base/test_base_checks.py:37]: should be in the cotext

Comment (JKL98ISR) [reply tests/vision/base/test_base_checks.py:37]: also use type hints

Comment (matanper) [reply tests/vision/base/test_base_checks.py:37]: why? it's for the tests

Comment (JKL98ISR) [reply tests/vision/base/test_base_checks.py:37]: oops meant conftest

**Comment by JKL98ISR** [comment]
also other functions missing type hints i.e get_index_in_dataset

**Pull request merged**